### DISCUSSION TITLE
Feature Request: Ability to download view-only videos from Google Drive

### DISCUSSION THREAD

**Discussion by stas-sazhaev** [aclap-dev/video-downloadhelper#383] [Help and Q&A]
### Your question

Hello! VDH no longer sees media for private video files uploaded to Google Drive - there was no problem previously. This feature has been actively used by me for several months.

<img width="1351" alt="image" src="https://github.com/aclap-dev/video-downloadhelper/assets/115798642/c06b42c6-a107-48e7-b7d7-cc07ff0d338e">

### Addon & CoApp version and details

version: 8.2.2.8
channel: stable
build date: 2024-03-18
build options: linuxlic: false, noyt: false, browser: firefox
lang: en-GB
license: accepted
key (only 16 first characters): f2baf3b9d7d04776
platform: aarch64 mac
UA: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:124.0) Gecko/20100101 Firefox/124.0
coapp: yes. Info:
{
  "id": "net.downloadhelper.coapp",
  "name": "VdhCoApp",
  "version": "2.0.19",
  "binary": "/Applications/net.downloadhelper.coapp.app/Contents/MacOS/vdhcoapp",
  "displayName": "VdhCoApp",
  "description": "Video DownloadHelper companion app",
  "target": {
    "os": "mac",
    "arch": "arm64",
    "node": "18"
  },
  "home": "/Users/stanislav",
  "converterBinary": "/Applications/net.downloadhelper.coapp.app/Contents/MacOS/ffmpeg",
  "converterBase": "ffmpeg",
  "converterBaseVersion": "n6.1-11-gc30f360955"
}

**Comment by Mattmjs** [comment]
Have you verified antivirus and firewall are
not blocking the companion app.

**Reply by stas-sazhaev** [reply]
Yes, sure.

**Comment by Mattmjs** [comment]
Ok, try going in Settings>More>Reset settings

If preferences is ticked, remove it. Give a tick to Video qualities then
click reset settings. See if this helps or not.

**Reply by stas-sazhaev** [reply]
Did a reset. Unfortunately, this didn't help. :(

**Comment by hdsghn** [comment]
Google Drive videos have also not been working for me.  Has been going on for a few months.  I booted up my old laptop which has a very old VDH version (7.something) and worked for Google Drive as of last August.  That no longer does which makes me suspect Google changed how they serve videos.  

I uploaded a test video to my own Drive account which VDH cannot see if that helps in troubleshooting: https://drive.google.com/file/d/1R2f3N2-O6t64Oe3HHuAMG70oTfkUfnsl/view

No errors, just does not see any media.  

version: 8.2.2.8
channel: stable
build date: 2024-03-18
build options: linuxlic: false, noyt: false, browser: firefox
lang: en-US
license: accepted
key (only 16 first characters): c3f21e38e9fcf775
platform: x86-64 win
UA: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:124.0) Gecko/20100101 Firefox/124.0
coapp: yes. Info:
{
  "id": "net.downloadhelper.coapp",
  "name": "VdhCoApp",
  "version": "2.0.19",
  "binary": "C:\\Program Files\\DownloadHelper CoApp\\vdhcoapp.exe",
  "displayName": "VdhCoApp",
  "description": "Video DownloadHelper companion app",
  "target": {
    "os": "windows",
    "arch": "x86_64",
    "node": "18"
  },
  "home": "C:\\Users\\micro",
  "converterBinary": "C:\\Program Files\\DownloadHelper CoApp\\ffmpeg.exe",
  "converterBase": "ffmpeg",
  "converterBaseVersion": "c30f360955"
}

**Comment by Mattmjs** [comment]
It does appear that drive videos are not being
detected. Going to need @paulrouget to look at that one. I still have detection on YouTube.

Are you getting detection elsewhere? 
Are you using YouTube premium?

**Reply by stas-sazhaev** [reply]
I use YouTube premium. Yes, I just checked and **I also have detection on YouTube.**

So the problem is now only relevant for video on the drive.

**Comment by stas-sazhaev** [comment]
Hello! It seems that VDH does not work with videos that were uploaded to Google Drive in 2024 (since the last Google update).

@Mattmjs @paulrouget 
Are there plans to improve the VDH so that Google Drive videos can again be downloaded?

**Reply by Mattmjs** [reply]
I'm sure it can be looked into and fixed but I'm a user like you. Paul does the coding for the software. 

I did a bit of digging around in the developer tools / network and found the streaming media in the Headers of a html subdocument 
for the domain youtube.googleapis.com
![Screenshot 2024-04-17 155129](https://github.com/aclap-dev/video-downloadhelper/assets/160297904/b59a669a-ca2e-4c47-a43e-cb0b04bec534)

I just right clicked on the highlighted area in the image and selected copy all and pasted in notepad. There are 3 video qualities on the moose video,480p,360p & 240p plus the audio. Using the find Ctrl + F in notepad look for 480p. It'll find the link for 480p that I've highlighted all the way up to the back slash. Then use the replace function to replace all instances of \\u0026  (there are 2 back slashes) with &

![drive](https://github.com/aclap-dev/video-downloadhelper/assets/160297904/6bb9bfd6-bb95-4593-b7e7-d8fc1082fce8)

Do the same process for the audio by looking for mp4a. You'll then have links for both video and audio by pasting them in new tabs. Then right click save video / audio as. Use VLC to play them together or merge them together using VDH, ffmpeg or any other
software. This shows where the media is, there may be HLS media somewhere which I think Paul can look into.

**Reply by stas-sazhaev** [reply]
Thanks a lot! I appreciate your response and your help.
Unfortunately, I couldn't repeat this trick. I don’t have similar links to media streaming in my network.
The domain youtube.googleapis.com finds only a post method request.

This is what it looks like...
<img width="1440" alt="example" src="https://github.com/aclap-dev/video-downloadhelper/assets/115798642/957715b0-207b-45ca-b5ca-790db5bdbca4">

**Reply by Mattmjs** [reply]
I'd be interested to have a look but if it's private, you could record the network activity and share it here as an attachment.

![3-10-23](https://github.com/aclap-dev/video-downloadhelper/assets/160297904/ae5db547-10b6-4ae9-a918-ec29d14ebbbf)

**Reply by stas-sazhaev** [reply]
@Mattmjs We can use this example video: https://drive.google.com/file/d/1R2f3N2-O6t64Oe3HHuAMG70oTfkUfnsl/view

Here is my HAR from this example.
[drive.google.com_Archive [24-04-18 19-59-20].har.zip](https://github.com/aclap-dev/video-downloadhelper/files/15028091/drive.google.com_Archive.24-04-18.19-59-20.har.zip)

**Reply by Mattmjs** [reply]
That was the example I already used to find the streaming media above, but I looked at your recorded network and it had
the media information on the same domain link. I don't know why it's different in your screenshot. 
All I've shown is that the streaming media is in there if you want to get it yourself or wait for an update in the extension
that detects drive video. You could also try another tool in **yt-dlp** which might work. Just give it the google drive link
and it should find media, at least on public google drive video it does. If it's private I don't know.

**Comment by EloyFarina** [comment]
Any workaround for this?

**Comment by GrampaWildWilly** [comment]
Refer to #414.

**Reply by stas-sazhaev** [reply]
Thank you for your reply. However, I’d like to emphasize once more that our discussion pertains specifically to downloading view-only videos, for which the Google Drive interface lacks a download function.

**Reply by GrampaWildWilly** [reply]
I didn't see this discussion until now.  The moose video is an example like what I was asking for in #414.  You are right.  There is no download button offered by Google Drive when it's view-only.  That sucks.

OK.  Matt's idea of looking at the headers of an unexpected object works, but there's a tweak you have to add.  Here's what I did.  In the Network Monitor, find an object of Type html with the word "embed" contained in its name.  Then copy all of the headers from that object.

![#01](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/18388809-2524-4717-8f8d-d3216ab87252)

Paste whatever that gives you into your text editor.  I use Notepad++ which helps a lot.  Find any occurrence of the string mimeType.  Notepad++ highlights every occurrence of a string once you highlight one such occurrence.

![#02](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/54796350-0389-4af6-80bb-81bbfc68e478)

Immediately following mimeType, it tells you what this ugly thing is describing.  A small distance before mimeType, it tells you the vertical component of the resolution of this particular item.  I've circled those indicators in the above image.  There are more mimeType descriptors below this section of the file, which you can see if you scroll.  I haven't bothered to scroll this . . . yet.  Let's get the URL in the 1080 descriptor.  The URL is a short way after the mimeType string.

![#03](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/ef06c309-6186-4b0d-aff4-af25e2b93a4d)

The URL is marked by ``\"`` (back slash, double quotation mark) at both the beginning & the end.  Everything within the pair of ``\"`` is the URL.  Copy/paste the URL out to another temporary file in your text editor.

![#04](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/02522951-2281-4a7c-b6b3-aa32fe55d736)

This URL contains something that is not valid in a URL.  Backslashes are not valid in any URL.  What I'm pointing out here is that the string I have highlighted needs to be changed.

![#05](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/97c7cd04-474a-45ba-aa49-a2cdcf77b500)

Every occurrence of ``\\u0026`` needs to be replaced with an ampersand.  Now you have a usable URL.  In Notepad++, I can just double click it.

![#06](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/01cf3634-d5f0-4127-a3f4-5ce9beadb346)

That causes a new browser window to open.  And look at that.  VDH can download it.  It will be video without audio.

Now, going back to the thing you pasted from the Network Monitor into your text editor, scroll it down a bit.  You will find the URL for the audio track.

![#07](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/b3b4af09-3d4a-455d-9653-85eb688819f9)

Do the same thing as before.  Copy/paste it elsewhere & replace all ``\\u0026`` with &.  Then double click it.  VDH can download this as well.

![#08](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/aa4117f6-ec90-4fe0-a899-b6c7c588e74e)

Oops.  Apparently not.

![#09](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/9dabce0c-42eb-4907-a132-c2e8f037f275)

Oh well.  No problem.  The browser has no problems downloading this.

![#10](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/27b44efe-e50c-4e37-888d-b693f77cd3f3)

Then just play the video & the audio in VLC using its synchronous playback feature.

![#11](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/7806e985-53c1-43b0-a47c-6885dec78ecc)

It turns out this audio track is actually empty.  But that's just for this case.  In general, I would assume you have videos that include audio.  Maybe the fact that it's empty is what tripped up VDH.  Maybe for another video that actually has audio, VDH might actually download the audio.  That would be for some future time.

Or you can merge the 2 tracks using VDH itself.  Oh wait.  It doesn't have that feature yet.  I think Paul is working on it.  OK.  Both ffmpeg & VLC have ways of doing it.  I'll leave that as an exercise for you.

**Reply by stas-sazhaev** [reply]
Thanks a lot for your help and this detailed guide! It's useful! My problem is that I need to download dozens of fairly long videos. Doing it the way described is going to be a bit tough. :-) That's why I bought a license for this plugin in the first place.

If @paulrouget  has a chance to fix this feature in the upcoming updates, I would be thrilled! It would help a lot.

**Reply by Mattmjs** [reply]
What if there was another method that was a bit easier, it would still require using the developer tools.
It would open the video and audio as separate files and then you'd need to use a software of your choice to merge the files
together. What would you think about that ?

**Reply by stas-sazhaev** [reply]
Yes, that would be helpful! Actually, it's not necessary to merge the video and audio. I can watch them using VLC player. The main issue is the number of steps you have to take to download them and the need to find the right piece of code in developer mode. The methods described above take a lot of time and attention. And when you need to download many separate videos, it becomes labor-intensive.

**Reply by Mattmjs** [reply]
Ok great you don't need to merge the files. I found a discussion on Reddit on this topic of Google Drive video.
Someone showed you could edit the streaming links shown as **videoplayback**  in the network. 
These are both the video and audio, if your not sure which is video or audio you can filter by **mime=video** & **mime=audio**.
Just copy the link and paste in notepad or any other text editor and chop out the part I show in the image.
Copy & paste again in a new tab and the media should play as well as downloading it.

![Screenshot 2024-07-20 225753](https://github.com/user-attachments/assets/488ec932-a8ff-4ce8-b73a-b26f62b76919)

**Reply by stas-sazhaev** [reply]
@Mattmjs, thanks a lot! I tried it — it works! And it is easy and quick.

I also noticed something interesting: when I open a new tab and save the file, the download speed is very low, around 50 kb/s, but if I close that tab, the speed increases significantly!

Thanks so much for the help! I'm grateful and appreciative.

PS I renamed this thread to a feature request, thinking that if this can be solved with such a small code tweak, we might expect this feature to appear in future plugin versions.

**Comment by Mattmjs** [comment]
I don't think I mentioned this before but try enabling low quality media and see if there is detection.

**Reply by GrampaWildWilly** [reply]
Good idea but made no difference.

**Comment by EloyFarina** [comment]
It worked!!!
Thanks for the workaround @GrampaWildWilly . I would like that I can perform the same easily with the tool.

**Reply by GrampaWildWilly** [reply]
Yeah, I think we all would like that.  But that's the way it is today.  It seems to be something Google did that makes it annoying.  They're not the only annoying site.  You can read about many other annoying web sites here in this forum.  But I'm glad to hear that the workaround works.  In your case, was it something that had a proper audio track?  Was VDH able to download the audio track, or did you have to resort to the browser's own Save Audio As function?

**Reply by EloyFarina** [reply]
In my case, I had to refresh the page multiples times, because I was not seeing the proper URL.
I had to paste the URL in notepad, not in Notepad++ because it was pasting all in 1 line.
I searched the last "quality":"hd720", I copied the URL in a tab and then the tool recognized the video as VIDEOPLAYBACK.MP4
I had to wait the same lenght of the video in order to finish to download.
When the video was downloaded, the audio track was included.

**Reply by GrampaWildWilly** [reply]
This wouldn't be something whose URL you could post here, is it?  The speed thing seems to be an issue but not uniformly.  I have seen Google (in the form of YouTube) give putrid download speed, only 200,000 or 300,000 bytes per second, even slower.  I have seen it give multi-million bytes per second download speed.  You just never know what you're going to get.  I would be curious to see what I get.  I would also be curious to know what happens if you retry it now, a few hours later.  Sometimes Save Video As in the browser runs faster than VDH, sometimes the other way around.

Pasting all in one line is what you get from these things.  Notepad++ tries to format some input like that, but you can see in my images above that large expanses of the text still just come out like amorphous blobs.

And the audio was in the file with the video . . .  That's interesting.  It must be your lucky day.

**Comment by Aplietexe** [comment]
[yt-dlp](https://pypi.org/project/yt-dlp/) seems to be able to download view-only videos from Drive, at least if they are publicly available to anyone with the link. Not sure if there is a way to use browser cookies with yt-dlp to download videos shared with specific accounts/organizations?
